File: javascript/frameworks/openai/agent.ts
/**
 * OpenAI-compatible agent with Scalekit-authenticated Gmail tools.
 * Run: npx tsx javascript/frameworks/openai/agent.ts
 */
import { ScalekitClient } from '@scalekit-sdk/node'
import { ConnectorStatus } from '@scalekit-sdk/node/lib/pkg/grpc/scalekit/v1/connected_accounts/connected_accounts_pb.js'
import OpenAI from 'openai'
import 'dotenv/config'

const scalekit = new ScalekitClient(
  process.env.SCALEKIT_ENVIRONMENT_URL!,
  process.env.SCALEKIT_CLIENT_ID!,
  process.env.SCALEKIT_CLIENT_SECRET!,
)

const openai = new OpenAI({
  baseURL: process.env.OPENAI_BASE_URL,
  apiKey: process.env.OPENAI_API_KEY,
})

const { connectedAccount } = await scalekit.actions.getOrCreateConnectedAccount({
  connectionName: 'gmail',
  identifier: 'user_123',
})
if (connectedAccount?.status !== ConnectorStatus.ACTIVE) {
  const { link } = await scalekit.actions.getAuthorizationLink({
    connectionName: 'gmail',
    identifier: 'user_123',
  })
  console.log('Authorize Gmail:', link)
  process.exit(0)
}

const { tools } = await scalekit.tools.listScopedTools('user_123', {
  filter: { connectionNames: ['gmail'] },
})
const llmTools: OpenAI.ChatCompletionTool[] = tools
  .map((tool) => tool.tool?.definition)
  .filter((definition): definition is NonNullable<typeof definition> => Boolean(definition?.name))
  .map((definition) => ({
    type: 'function',
    function: {
      name: String(definition.name),
      description: String(definition.description ?? ''),
      parameters: (definition.input_schema ?? { type: 'object', properties: {} }) as Record<string, unknown>,
    },
  }))

const messages: OpenAI.ChatCompletionMessageParam[] = [
  { role: 'user', content: 'Fetch my last 5 unread emails and summarize them' },
]

while (true) {
  const response = await openai.chat.completions.create({
    model: process.env.OPENAI_MODEL || 'claude-sonnet-4-6',
    tools: llmTools,
    messages,
  })
  const message = response.choices[0].message
  if (!message.tool_calls?.length) {
    console.log(message.content)
    break
  }

  messages.push(message)
  for (const toolCall of message.tool_calls) {
    if (!('function' in toolCall)) {
      continue
    }
    const result = await scalekit.actions.executeTool({
      toolName: toolCall.function.name,
      identifier: 'user_123',
      toolInput: JSON.parse(toolCall.function.arguments),
    })
    messages.push({ role: 'tool', tool_call_id: toolCall.id, content: JSON.stringify(result.data) })
  }
}

File: python/frameworks/openai/agent.py
"""
OpenAI-compatible agent with Scalekit-authenticated Gmail tools.
Run: python python/frameworks/openai/agent.py
"""
import json
import os
import scalekit.client
from dotenv import find_dotenv, load_dotenv
from google.protobuf.json_format import MessageToDict
from openai import OpenAI

load_dotenv(find_dotenv())

scalekit_client = scalekit.client.ScalekitClient(
    client_id=os.getenv("SCALEKIT_CLIENT_ID"),
    client_secret=os.getenv("SCALEKIT_CLIENT_SECRET"),
    env_url=os.getenv("SCALEKIT_ENVIRONMENT_URL"),
)
actions = scalekit_client.actions
client = OpenAI(base_url=os.getenv("OPENAI_BASE_URL"), api_key=os.getenv("OPENAI_API_KEY"))

response = actions.get_or_create_connected_account(
    connection_name="gmail",
    identifier="user_123",
)
if response.connected_account.status != "ACTIVE":
    link = actions.get_authorization_link(connection_name="gmail", identifier="user_123")
    print("Authorize Gmail:", link.link)
    input("Press Enter after authorizing...")

scoped_response, _ = actions.tools.list_scoped_tools(
    identifier="user_123",
    filter={"connection_names": ["gmail"]},
)
llm_tools = [
    {
        "type": "function",
        "function": {
            "name": MessageToDict(tool.tool).get("definition", {}).get("name"),
            "description": MessageToDict(tool.tool).get("definition", {}).get("description", ""),
            "parameters": MessageToDict(tool.tool).get("definition", {}).get("input_schema", {}),
        },
    }
    for tool in scoped_response.tools
]

messages = [{"role": "user", "content": "Fetch my last 5 unread emails and summarize them"}]

while True:
    response = client.chat.completions.create(
        model=os.getenv("OPENAI_MODEL", "claude-sonnet-4-6"),
        tools=llm_tools,
        messages=messages,
    )
    message = response.choices[0].message
    if not message.tool_calls:
        print(message.content)
        break

    messages.append(message)
    for tool_call in message.tool_calls:
        result = actions.execute_tool(
            tool_name=tool_call.function.name,
            identifier="user_123",
            tool_input=json.loads(tool_call.function.arguments),
        )
        messages.append(
            {
                "role": "tool",
                "tool_call_id": tool_call.id,
                "content": str(result.data),
            }
        )
